Linewrap PR description to 72 chars automatically

Clyybber opened the pull request:
## Summary
* Linewrap the PR description to 72 chars automatically when merging a
PR
* Codeblocks (```` ``` ````, `~~~` seperated or indented with 4 or more spaces, see
https://github.github.com/gfm/#indented-code-block ) are not linewrapped,
and codespans `like this` are treated as single words

## Details
* Remove the manual 72 char requirement and merge the top comment into
the comment below the section break
* Restrict the removal of html comment introduced in
https://github.com/nim-works/nimskull/commit/a862810f939f6f45ef8c54a135f39cfa060875a1
to non-codeblocks

Clyybber committed:
Text wrap PR descriptions

Clyybber committed:
Remove 72 chars line length requirement

and move the comment below the section break

alaviss commented on .github/workflows/slash-command-merge.yml:
```diff
@@ -51,8 +51,50 @@ jobs:
+            function getSegments(rex, str) {
+              var segments = [];
+              var lastIndex = 0;
+              var match;
```
Avoid `var` please, that stuff defines a global that is reachable from any scope.

Move the declarations outside if you need them preserved.

alaviss commented on .github/workflows/slash-command-merge.yml:
````diff
@@ -51,8 +51,50 @@ jobs:
+              getSegments(/(?:^ {0,3}(```|~~~)[^]*?^ {0,3}\1 *$)|^(?:\n {4,}.*)+/gm, text)
+                .map( ([isCodeBlock, segment]) =>
+                  isCodeBlock
+                  ? segment // Don't touch code blocks
+                  : segment.split('\n').map(line =>
+                      line.length <= maxLineLength
+                      ? line // Line isn't too long, nothing to do
+                      : getSegments(/(`+).*?\1/g, line)
+                          // Treat codespans as single words:
+                          .flatMap( ([isCodeSpan, segment]) => isCodeSpan ? segment : segment.split(' ') )
+                          .reduce(([newLine, lineLength], word) =>
+                            newLine
+                            // Don't touch trailing spaces:
+                            ? word.length > 0 && lineLength + 1 + word.length > maxLineLength
+                              ? [newLine + '\n' + word, word.length]
+                              : [newLine + ' ' + word, lineLength + 1 + word.length]
+                            : [newLine + word, lineLength + word.length]
````
This is horrible code, sorry.

Use `filter` to remove the excessive ternary. Performance doesn't matter much over here.

Clyybber replied:
Lol, yeah I figured nested ternaries would be kinda hard to read. Blame it on js for not having if-expressions :p
Started out with just for loops and mostly imperative code and slowly converted it into this.
I don't see how you would replace these with filter though, maybe I'm missing something.

alaviss replied:
Oh true, sorry.

You can write those as like actual functions instead of chaining expressions :P

saem replied:
I like the ternaries, the shape of the code fits the logic of it. Bouncing up and down between function definitions and the logic doesn't seem like a win at all.

saem replied:
I usually put the `?` and `:` as trailing on the line previous, so the reader knows they're entering into a ternary before leaving the current line. Helps reading flow, IMO.

Clyybber replied:
Because of the nesting in this case, them trailing the previous line looks a bit confusing IMO:
```js
newLine ?
  word.length > 0 && lineLength + 1 + word.length > maxLineLength ?
    [newLine + '\n' + word, word.length] :
    [newLine + ' ' + word, lineLength + 1 + word.length] :
  [newLine + word, lineLength + word.length]
```

alaviss reviewed (commented):
Some initial thoughts.

And I think we should move this out into its own file. I think inlining in YAML stops making sense once the script gets complex.

Clyybber committed:
Use let instead of var in getSegments

Clyybber committed:
Extract into tools/ci_format_pr_description.js and add tests

Clyybber commented on tools/koch/koch.nim:
```diff
@@ -24,6 +24,7 @@ import std/[json, os, strutils, parseopt, osproc, times]
   # Using `std/os` instead of `os` to fail early if config isn't set up properly.
   # If this fails with: `Error: cannot open file: std/os`, see
   # https://github.com/nim-lang/Nim/pull/14291 for explanation + how to fix.
+import ../compiler/utils/nodejs
```
Maybe it's better to put the tests in a `ci` category run by testament, as testament already imports compiler/utils/nodejs anyways..

saem replied:
IMO testament `ci` category would be better, it keeps `koch` that much slimmer.

Clyybber replied:
I put it into the `tests/tools` category, since the javascript file also sits under `tools`.

Clyybber committed:
Rename newLine to wrappedLine

Because newLine kind of implies it being a single
line which is not the case if it was too long

Clyybber committed:
Actually use node executable found with findNodeJs

Clyybber committed:
Remove unneccessary newline in PR template

saem commented on tools/ci_format_pr_description.js:
```diff
@@ -0,0 +1,192 @@
+              .reduce(([wrappedLine, lineLength], word) =>
+                wrappedLine
+                // Don't touch trailing spaces:
+                ? word.length > 0 && lineLength + 1 + word.length > maxLineLength
+                  ? [wrappedLine + '\n' + word, word.length]
+                  : [wrappedLine + ' ' + word, lineLength + 1 + word.length]
```
did you consider assigning this to a temporary?

admittedly, it's slightly less expression-y.

Clyybber replied:
Yeah, it would force one to use if statements with returns inside instead. Found it cleaner this way.

saem reviewed (approved):
LGTM

My comments are non-blocking, happy to discuss further if you'd like.

zerbina commented:
I think the HTML comment stripping should stay in some form, or at least be reinstated again later. Without it, or something to take its place, each PR message would have to manually be inspected for possible HTML comments.

saem commented:
> I think the HTML comment stripping should stay in some form, or at least be reinstated again later. Without it, or something to take its place, each PR message would have to manually be inspected for possible HTML comments.

Is that blocking for this PR?

alaviss commented:
I think this PR is a good first step.

If we want to increase the feature set in the future, we can spin this workflow out into an action repo where we can even pull in a markdown parser to handle comments and such properly.

Clyybber committed:
Don't dedent lines and move tests to tools category

Clyybber committed:
Remove HTML comments too

Also improve formatting of the code in this PR

Clyybber commented:
I readded HTML comment removal outside of codeblocks and moved the test into the `tools` category. Should be ready to go.

Clyybber committed:
Rename to formatPRdescription

now that comments are stripped too
and add some documentation

saem commented on tests/tools/tci_forrmat_pr_description.nim:
```diff
@@ -0,0 +1,169 @@
+target: js
+"""
+
+import jsffi
```
```suggestion
import std/jsffi
```

saem replied:
Not a hard blocker

saem reviewed (approved):
One minor suggestion, nonblocking. Otherwise it's got to go.

Clyybber committed:
Use std/ prefxed import in the test

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

Clyybber commented:
Thanks for the reviews!

Clyybber commented:
/merge

Clyybber committed:
Fix typo in test filename and prevent joining

as joining with megatest will cause the relative import to break
